### PR TITLE
Extract tag authorization in a dedicated service

### PR DESCRIPTION
**Pull request by juherr**
The objective of the change is to be able to delegate the authorization responsibility to an outside EMSP service if wanted.

WDYT?

### CODE REVIEW

**Comment by goekay** [comment]
late feedback:

i think the goal is valuable. couple of suggestion from my side:

- can we do these without database changes?
- your changes are only encapsulating the auth part while leaving `OcppTagActivityRecord record = ocppTagRepository.getRecord(idTag);` as is. i would make this also part of the auth, because in such external cases the record does not necessarily have to be on local db.

**Comment by juherr** [comment]
@goekay done as suggested

File: src/main/java/de/rwth/idsg/steve/service/AuthTagService.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+@RequiredArgsConstructor
+public class AuthTagService {
+  private final OcppTagRepository ocppTagRepository;
+  private final SettingsRepository settingsRepository;
+
+  public IdTagInfo decideStatus(String idTag,
+                                boolean isStartTransactionReqContext,
+                                @Nullable String chargeBoxId,
+                                @Nullable Integer connectorId) {
+    OcppTagActivityRecord record = ocppTagRepository.getRecord(idTag);
```
Comment (goekay) [inline src/main/java/de/rwth/idsg/steve/service/AuthTagService.java:29]: the project uses 4-spaces indentation

note to self: we should enforce this by checkstyle.

File: src/main/java/de/rwth/idsg/steve/service/OcppTagService.java [hunk 2]
```diff
@@ -97,35 +98,36 @@ public void removeUnknown(List<String> idTagList) {
     }
 
     @Nullable
-    public IdTagInfo getIdTagInfo(@Nullable String idTag, boolean isStartTransactionReqContext) {
+    public IdTagInfo getIdTagInfo(@Nullable String idTag, boolean isStartTransactionReqContext,
+        @Nullable String chargeBoxId, @Nullable Integer connectorId) {
```
Comment (goekay) [inline src/main/java/de/rwth/idsg/steve/service/OcppTagService.java:102]: method parameter alignment after linebreak not consistent.

File: src/main/java/de/rwth/idsg/steve/service/OcppTagService.java [hunk 3]
```diff
@@ -97,35 +98,36 @@ public void removeUnknown(List<String> idTagList) {
...
+                return idTagInfo;
 
             case BLOCKED:
             case EXPIRED:
             case CONCURRENT_TX:
             case ACCEPTED:
-                return new IdTagInfo().withStatus(status)
-                                      .withParentIdTag(record.getParentIdTag())
-                                      .withExpiryDate(getExpiryDateOrDefault(record));
+                return idTagInfo;
```
Comment (goekay) [inline src/main/java/de/rwth/idsg/steve/service/OcppTagService.java:118]: we might question whether we need all the switch cases here (i think not). i guess we can just can use

```
if (idTagInfo.getStatus() == INVALID) {
    invalidOcppTagService.processNewUnidentified(idTag);
}

return idTagInfo
```

File: src/main/java/de/rwth/idsg/steve/service/OcppTagService.java [hunk 4]
```diff
@@ -97,35 +98,36 @@ public void removeUnknown(List<String> idTagList) {
...
             default:
                 throw new SteveException("Unexpected AuthorizationStatus");
         }
     }
 
     @Nullable
-    public IdTagInfo getIdTagInfo(@Nullable String idTag, boolean isStartTransactionReqContext, Supplier<IdTagInfo> supplierWhenException) {
+    public IdTagInfo getIdTagInfo(@Nullable String idTag, boolean isStartTransactionReqContext,
+        @Nullable String chargeBoxId, @Nullable Integer connectorId,
+        Supplier<IdTagInfo> supplierWhenException) {
```
Comment (goekay) [inline src/main/java/de/rwth/idsg/steve/service/OcppTagService.java:128]: method parameter alignment after linebreak not consistent.

File: src/main/java/de/rwth/idsg/steve/utils/OcppTagActivityRecordUtils.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+import jooq.steve.db.tables.records.OcppTagActivityRecord;
+import lombok.AccessLevel;
+import lombok.NoArgsConstructor;
+import org.joda.time.DateTime;
+
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public final class OcppTagActivityRecordUtils {
+
+  public static boolean isExpired(OcppTagActivityRecord record, DateTime now) {
```
Comment (goekay) [inline src/main/java/de/rwth/idsg/steve/utils/OcppTagActivityRecordUtils.java:11]: the project uses 4-spaces indentation

**Review by goekay — CHANGES_REQUESTED** [review]

**Comment by goekay** [comment]
@juherr are you willing to do the changes or should i do them after merging?

**Comment by juherr** [comment]
@goekay I'm currently in vacations and without a computer for the next 2 weeks.

You can wait or make the changed.
Fyi, you should be able to push on my branch and fix before the merge.

**Review by goekay — APPROVED** [review]

**Comment by goekay** [comment]
thanks @juherr !

**Comment by juherr** [comment]
@goekay thanks for the merge. Please review the other pull requests if you find some time

**Pull request merged**